File: src/commands/engine.ts
/**
 * Engine command for ubuild CLI
 *
 * Displays information about the Unreal Engine installation
 * associated with the current project or specified path.
 *
 * @module commands/engine
 */

import { Command } from 'commander';
import chalk from 'chalk';
import { Writable } from 'stream';
import { EngineResolver } from '../core/engine-resolver';
import { Logger } from '../utils/logger';
import { ProjectDetector } from '../core/project-detector';
import { handleCommandError } from '../utils/error';

/** Options for the engine command. */
export interface EngineCommandOptions {
  /** Path to project directory or .uproject file */
  project?: string;
  /** Output result as JSON */
  json?: boolean;
  /** Show verbose engine detection details */
  verbose?: boolean;
  /** Writable stream for standard output (defaults to process.stdout) */
  stdout?: Writable;
  /** Writable stream for error output (defaults to process.stderr) */
  stderr?: Writable;
}

/**
 * Executes the engine command to display engine information.
 *
 * Resolves the Unreal Engine installation for the current project and displays
 * detailed information about the engine including version, path, and association.
 *
 * @param options - Configuration options for the engine command
 * @returns Promise that resolves when the command completes
 *
 * @example
 * ```typescript
 * await executeEngine({ project: './MyProject', verbose: true });
 * ```
 */
export async function executeEngine(options: EngineCommandOptions): Promise<void> {
  const stdout = options.stdout || process.stdout;
  const stderr = options.stderr || process.stderr;

  const logger = new Logger({ stdout, stderr });

  try {
    if (!options.json) {
      logger.title('Engine Information');
    }

    let projectPath = options.project || process.cwd();

    const projectResult = await ProjectDetector.detectProject({ cwd: projectPath });
    if (projectResult.project) {
      projectPath = projectResult.project.path;
    }

    const result = await EngineResolver.resolveEngine(projectPath);

    if (options.json) {
      logger.json(result);
      return;
    }

    if (options.verbose) {
      logger.subTitle('Engine Detection Details');
      const allInstallations = await EngineResolver.findEngineInstallations();
      logger.write(`Total engines detected: ${allInstallations.length}`);

      if (allInstallations.length > 0) {
        allInstallations.forEach((engine, index) => {
          logger.write(`\n  Engine ${index + 1}:`);
          logger.write(`    Path: ${engine.path}`);
          logger.write(`    Source: ${engine.source || 'unknown'}`);
          logger.write(`    Association ID: ${engine.associationId}`);
          logger.write(`    Display Name: ${engine.displayName || '(none)'}`);
          if (engine.version) {
            logger.write(
              `    Version: ${engine.version.MajorVersion}.${engine.version.MinorVersion}.${engine.version.PatchVersion}`
            );
          }
          if (engine.installedDate) {
            logger.write(`    Installed: ${engine.installedDate}`);
          }
        });
      }
      logger.write('\n');
    }

    if (result.error) {
      logger.error(result.error);
      throw new Error(result.error);
    }

    if (result.engine && result.uprojectEngine) {
      const engine = result.engine;
      logger.success(
        `Found engine for project: ${chalk.bold(engine.displayName || engine.associationId)}`
      );

      logger.subTitle('Engine Details');
      logger.write(`  Path: ${engine.path}`);

      if (engine.version) {
        logger.write(
          `  Version: ${engine.version.MajorVersion}.${engine.version.MinorVersion}.${engine.version.PatchVersion}`
        );
        logger.write(`  Build ID: ${engine.version.BuildId}`);
        logger.write(`  Branch: ${engine.version.BranchName}`);
        logger.write(`  Changelist: ${engine.version.Changelist}`);
        logger.write(`  Promoted Build: ${engine.version.IsPromotedBuild ? 'Yes' : 'No'}`);
      }

      logger.write(`  Association ID: ${engine.associationId}`);
      if (engine.installedDate) {
        logger.write(`  Installed: ${engine.installedDate}`);
      }
    } else if (!result.engine) {
      logger.warning('No engine installation found');
    }

    if (result.uprojectEngine) {
      logger.subTitle('Project Engine Association');
      logger.write(`  GUID: ${result.uprojectEngine.guid}`);
      if (result.uprojectEngine.name) {
        logger.write(`  Name: ${result.uprojectEngine.name}`);
      }
      if (result.uprojectEngine.path) {
        logger.write(`  Path: ${result.uprojectEngine.path}`);
      }
      if (result.uprojectEngine.version) {
        logger.write(`  Version: ${result.uprojectEngine.version}`);
      }

      if (!result.engine) {
        logger.warning(
          'Engine association found in project, but no matching engine installation detected'
        );
      }
    }

    if (result.warnings.length > 0) {
      logger.subTitle('Warnings');
      result.warnings.forEach((warning) => {
        logger.write(`  • ${warning}`);
      });
    }

    logger.write('\n');
    if (result.engine && result.uprojectEngine) {
      logger.success('Engine information retrieved successfully');
    } else if (!result.engine) {
      logger.warning('No engine installation found');
    }
  } catch (error) {
    throw error;
  }
}

/**
 * Registers the 'engine' command for displaying engine information.
 *
 * This command detects and displays information about the Unreal Engine
 * installation associated with the current project, including version,
 * installation path, and engine association details.
 *
 * @param program - The Commander program instance
 *
 * @example
 * ```typescript
 * engineCommand(program);
 * ```
 */
export function engineCommand(program: Command): void {
  program
    .command('engine')
    .description('Display engine information for the current project')
    .option('-p, --project <path>', 'Path to project directory or .uproject file')
    .option('-j, --json', 'Output result as JSON')
    .option('-v, --verbose', 'Show verbose engine detection details')
    .action(async (options) => {
      try {
        await executeEngine({
          project: options.project,
          json: options.json,
          verbose: options.verbose,
        });
      } catch (error) {
        handleCommandError(error, 'Engine command failed');
      }
    });
}

File: src/commands/engine.rs
use anyhow::Result;

use crate::cli::EngineArgs;
use crate::core::engine_resolver::EngineResolver;
use crate::core::project_path_resolver::ProjectPathResolver;
use crate::utils::logger::Logger;

pub fn execute(args: EngineArgs) -> Result<()> {
    Logger::title("Engine Information");

    let project_path = ProjectPathResolver::resolve_or_throw(args.project.as_deref()).ok();
    let result = EngineResolver::resolve_engine(project_path.as_deref());

    if args.json {
        let json = serde_json::json!({
            "engine": result.engine.as_ref().map(|e| serde_json::json!({
                "path": e.path,
                "associationId": e.association_id,
                "displayName": e.display_name,
                "version": e.version.as_ref().map(std::string::ToString::to_string),
                "source": e.source.to_string(),
            })),
            "uprojectEngine": result.uproject_engine.as_ref().map(|a| serde_json::json!({
                "id": a.id,
            })),
            "warnings": result.warnings,
        });
        println!("{}", serde_json::to_string_pretty(&json)?);
        return Ok(());
    }

    if let Some(ref engine) = result.engine {
        Logger::subtitle("Engine Details");
        Logger::info(&format!("Path: {}", engine.path.display()));
        if let Some(ref v) = engine.version {
            Logger::info(&format!("Version: {v}"));
            if !v.build_id.is_empty() {
                Logger::info(&format!("Build ID: {}", v.build_id));
            }
            if !v.branch_name.is_empty() {
                Logger::info(&format!("Branch: {}", v.branch_name));
            }
            Logger::info(&format!("Changelist: {}", v.changelist));
            Logger::info(&format!(
                "Promoted Build: {}",
                if v.is_promoted_build != 0 { "Yes" } else { "No" }
            ));
        }
        Logger::info(&format!("Association ID: {}", engine.association_id));
        if let Some(ref date) = engine.installed_date {
            Logger::info(&format!("Installed: {date}"));
        }
        println!();
    } else {
        Logger::warning("No engine installation detected");
    }

    if let Some(ref assoc) = result.uproject_engine {
        Logger::subtitle("Project Engine Association");
        Logger::info(&format!("ID: {}", assoc.id));
        println!();
    }

    if args.verbose {
        let installations = EngineResolver::find_engine_installations();
        if !installations.is_empty() {
            Logger::subtitle("Engine Detection Details");
            Logger::info(&format!("Total engines detected: {}", installations.len()));
            for (i, inst) in installations.iter().enumerate() {
                Logger::writeln(&format!("  Engine {}:", i + 1));
                Logger::writeln(&format!("    Path: {}", inst.path.display()));
                Logger::writeln(&format!("    Source: {}", inst.source));
                Logger::writeln(&format!("    Association ID: {}", inst.association_id));
                Logger::writeln(&format!("    Display Name: {}", inst.display_name));
                if let Some(ref v) = inst.version {
                    Logger::writeln(&format!("    Version: {v}"));
                }
                if let Some(ref d) = inst.installed_date {
                    Logger::writeln(&format!("    Installed: {d}"));
                }
            }
        }
    }

    for w in &result.warnings {
        Logger::warning(w);
    }

    Ok(())
}
